File: C#/qr_code_utils/qr_code_utils_test.cs
using System;
using AllToolkit;

namespace QRCodeUtilsTest
{
    /// <summary>
    /// Test suite for QRCodeUtils
    /// </summary>
    class Program
    {
        static int testsPassed = 0;
        static int testsFailed = 0;

        static void Main(string[] args)
        {
            Console.WriteLine("=== QR Code Utils Test Suite ===\n");

            TestBasicGeneration();
            TestTextGeneration();
            TestSvgGeneration();
            TestModeDetection();
            TestCapacityCalculation();
            TestErrorLevels();
            TestValidation();
            TestCapacityInfo();
            TestErrorLevelConversion();
            TestQRCodeSize();

            Console.WriteLine("\n=== Test Summary ===");
            Console.WriteLine($"Passed: {testsPassed}");
            Console.WriteLine($"Failed: {testsFailed}");
            Console.WriteLine($"Total: {testsPassed + testsFailed}");

            Environment.Exit(testsFailed > 0 ? 1 : 0);
        }

        static void TestBasicGeneration()
        {
            Console.WriteLine("Test: Basic Generation");
            try
            {
                var qr = QRCodeUtils.Generate("HELLO");
                Assert(qr != null, "QR Code should not be null");
                Assert(qr.Version >= 1 && qr.Version <= 10, "Version should be between 1 and 10");
                Assert(qr.Width == 17 + 4 * qr.Version, "Width should match version formula");
                Assert(qr.Matrix != null, "Matrix should not be null");
                Assert(qr.Matrix.GetLength(0) == qr.Width, "Matrix height should match width");
                Assert(qr.Matrix.GetLength(1) == qr.Width, "Matrix width should match width");
                Assert(qr.Data == "HELLO", "Data should match input");
                Console.WriteLine("  PASSED");
                testsPassed++;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  FAILED: {ex.Message}");
                testsFailed++;
            }
        }

        static void TestTextGeneration()
        {
            Console.WriteLine("Test: Text Generation");
            try
            {
                string text = QRCodeUtils.GenerateText("TEST", "##", "  ");
                Assert(!string.IsNullOrEmpty(text), "Text should not be empty");
                Assert(text.Contains("##"), "Text should contain dark modules");
                Assert(text.Contains("  "), "Text should contain light modules");
                Console.WriteLine("  PASSED");
                testsPassed++;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  FAILED: {ex.Message}");
                testsFailed++;
            }
        }

        static void TestSvgGeneration()
        {
            Console.WriteLine("Test: SVG Generation");
            try
            {
                string svg = QRCodeUtils.GenerateSvg("SVG TEST", 4);
                Assert(!string.IsNullOrEmpty(svg), "SVG should not be empty");
                Assert(svg.Contains("<?xml"), "SVG should contain XML declaration");
                Assert(svg.Contains("<svg"), "SVG should contain svg element");
                Assert(svg.Contains("</svg>"), "SVG should contain closing svg tag");
                Assert(svg.Contains("rect"), "SVG should contain rect elements");
                Console.WriteLine("  PASSED");
                testsPassed++;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  FAILED: {ex.Message}");
                testsFailed++;
            }
        }

        static void TestModeDetection()
        {
            Console.WriteLine("Test: Mode Detection");
            try
            {
                // Numeric mode
                var mode = QRCodeUtils.GetMode("12345");
                Assert(mode == QRMode.Numeric, "12345 should be Numeric mode");

                // Alphanumeric mode
                mode = QRCodeUtils.GetMode("HELLO");
                Assert(mode == QRMode.Alphanumeric, "HELLO should be Alphanumeric mode");

                // Byte mode
                mode = QRCodeUtils.GetMode("hello");
                Assert(mode == QRMode.Byte, "hello should be Byte mode");

                mode = QRCodeUtils.GetMode("Test123!");
                Assert(mode == QRMode.Byte, "Test123! should be Byte mode");

                Console.WriteLine("  PASSED");
                testsPassed++;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  FAILED: {ex.Message}");
                testsFailed++;
            }
        }

        static void TestCapacityCalculation()
        {
            Console.WriteLine("Test: Capacity Calculation");
            try
            {
                // Numeric capacity
                int capacity = QRCodeUtils.GetMaxCapacity(1, QRMode.Numeric, QRErrorCorrectionLevel.L);
                Assert(capacity > 0, "Numeric capacity should be positive");

                // Alphanumeric capacity
                capacity = QRCodeUtils.GetMaxCapacity(1, QRMode.Alphanumeric, QRErrorCorrectionLevel.L);
                Assert(capacity > 0, "Alphanumeric capacity should be positive");

                // Byte capacity
                capacity = QRCodeUtils.GetMaxCapacity(1, QRMode.Byte, QRErrorCorrectionLevel.L);
                Assert(capacity > 0, "Byte capacity should be positive");

                // Error level affects capacity
                int capL = QRCodeUtils.GetMaxCapacity(2, QRMode.Byte, QRErrorCorrectionLevel.L);
                int capH = QRCodeUtils.GetMaxCapacity(2, QRMode.Byte, QRErrorCorrectionLevel.H);
                Assert(capL > capH, "L level should have higher capacity than H level");

                Console.WriteLine("  PASSED");
                testsPassed++;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  FAILED: {ex.Message}");
                testsFailed++;
            }
        }

        static void TestErrorLevels()
        {
            Console.WriteLine("Test: Error Levels");
            try
            {
                var qrL = QRCodeUtils.Generate("TEST", QRErrorCorrectionLevel.L);
                Assert(qrL.ErrorCorrectionLevel == QRErrorCorrectionLevel.L, "Should use L level");

                var qrM = QRCodeUtils.Generate("TEST", QRErrorCorrectionLevel.M);
                Assert(qrM.ErrorCorrectionLevel == QRErrorCorrectionLevel.M, "Should use M level");

                var qrQ = QRCodeUtils.Generate("TEST", QRErrorCorrectionLevel.Q);
                Assert(qrQ.ErrorCorrectionLevel == QRErrorCorrectionLevel.Q, "Should use Q level");

                var qrH = QRCodeUtils.Generate("TEST", QRErrorCorrectionLevel.H);
                Assert(qrH.ErrorCorrectionLevel == QRErrorCorrectionLevel.H, "Should use H level");

                Console.WriteLine("  PASSED");
                testsPassed++;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  FAILED: {ex.Message}");
                testsFailed++;
            }
        }

        static void TestValidation()
        {
            Console.WriteLine("Test: Validation");
            try
            {
                // Valid data
                Assert(QRCodeUtils.ValidateData("TEST"), "TEST should be valid");
                Assert(QRCodeUtils.ValidateData("A"), "Single char should be valid");

                // Empty/null data should throw
                try
                {
                    QRCodeUtils.Generate("");
                    Assert(false, "Empty string should throw exception");
                }
                catch (ArgumentException)
                {
                    // Expected
                }

                try
                {
                    QRCodeUtils.Generate(null);
                    Assert(false, "Null should throw exception");
                }
                catch (ArgumentException)
                {
                    // Expected
                }

                Console.WriteLine("  PASSED");
                testsPassed++;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  FAILED: {ex.Message}");
                testsFailed++;
            }
        }

        static void TestCapacityInfo()
        {
            Console.WriteLine("Test: Capacity Info");
            try
            {
                var info = QRCodeUtils.GetCapacityInfo("HELLO WORLD");
                Assert(info != null, "Capacity info should not be null");
                Assert(info.DataLength == 11, "Data length should be 11");
                Assert(info.Mode == QRMode.Alphanumeric, "Mode should be Alphanumeric");
                Assert(info.Version >= 1, "Version should be at least 1");
                Assert(info.MaxCapacity > 0, "Max capacity should be positive");
                Assert(info.CanEncode, "Should be able to encode");

                // Test ToString
                string str = info.ToString();
                Assert(str.Contains("Mode:"), "ToString should contain Mode");
                Assert(str.Contains("Version:"), "ToString should contain Version");

                Console.WriteLine("  PASSED");
                testsPassed++;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  FAILED: {ex.Message}");
                testsFailed++;
            }
        }

        static void TestErrorLevelConversion()
        {
            Console.WriteLine("Test: Error Level Conversion");
            try
            {
                // To string
                Assert(QRCodeUtils.ErrorLevelToString(QRErrorCorrectionLevel.L) == "L", "L should convert to 'L'");
                Assert(QRCodeUtils.ErrorLevelToString(QRErrorCorrectionLevel.M) == "M", "M should convert to 'M'");
                Assert(QRCodeUtils.ErrorLevelToString(QRErrorCorrectionLevel.Q) == "Q", "Q should convert to 'Q'");
                Assert(QRCodeUtils.ErrorLevelToString(QRErrorCorrectionLevel.H) == "H", "H should convert to 'H'");

                // From string
                Assert(QRCodeUtils.StringToErrorLevel("L") == QRErrorCorrectionLevel.L, "'L' should parse to L");
                Assert(QRCodeUtils.StringToErrorLevel("M") == QRErrorCorrectionLevel.M, "'M' should parse to M");
                Assert(QRCodeUtils.StringToErrorLevel("Q") == QRErrorCorrectionLevel.Q, "'Q' should parse to Q");
                Assert(QRCodeUtils.StringToErrorLevel("H") == QRErrorCorrectionLevel.H, "'H' should parse to H");
                Assert(QRCodeUtils.StringToErrorLevel("invalid") == QRErrorCorrectionLevel.M, "Invalid should default to M");
                Assert(QRCodeUtils.StringToErrorLevel(null) == QRErrorCorrectionLevel.M, "Null should default to M");

                // Case insensitive
                Assert(QRCodeUtils.StringToErrorLevel("l") == QRErrorCorrectionLevel.L, "'l' should parse to L");
                Assert(QRCodeUtils.StringToErrorLevel("m") == QRErrorCorrectionLevel.M, "'m' should parse to M");

                Console.WriteLine("  PASSED");
                testsPassed++;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  FAILED: {ex.Message}");
                testsFailed++;
            }
        }

        static void TestQRCodeSize()
        {
            Console.WriteLine("Test: QR Code Size");
            try
            {
                Assert(QRCodeUtils.GetQRCodeSize(1) == 21, "Version 1 should be 21x21");
                Assert(QRCodeUtils.GetQRCodeSize(2) == 25, "Version 2 should be 25x25");
                Assert(QRCodeUtils.GetQRCodeSize(5) == 37, "Version 5 should be 37x37");
                Assert(QRCodeUtils.GetQRCodeSize(10) == 57, "Version 10 should be 57x57");

                try
                {
                    QRCodeUtils.GetQRCodeSize(0);
                    Assert(false, "Version 0 should throw");
                }
                catch (ArgumentException)
                {
                    // Expected
                }

                try
                {
                    QRCodeUtils.GetQRCodeSize(41);
                    Assert(false, "Version 41 should throw");
                }
                catch (ArgumentException)
                {
                    // Expected
                }

                Console.WriteLine("  PASSED");
                testsPassed++;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  FAILED: {ex.Message}");
                testsFailed++;
            }
        }

        static void Assert(bool condition, string message)
        {
            if (!condition)
            {
                throw new Exception($"Assertion failed: {message}");
            }
        }
    }
}

File: Java/qr_code_utils/qr_code_utils_test.java
package qr_code_utils;

import java.awt.image.BufferedImage;

import static qr_code_utils.mod.*;

/**
 * QR Code Utilities Test Suite
 * 
 * Comprehensive tests for QR code generation functionality.
 * Tests cover all encoding modes, error correction levels, and output formats.
 * 
 * @author AllToolkit Contributors
 * @version 1.0.0
 */
public class qr_code_utils_test {
    
    private static int testsPassed = 0;
    private static int testsFailed = 0;
    
    public static void main(String[] args) {
        System.out.println("========================================");
        System.out.println("QR Code Utilities Test Suite");
        System.out.println("========================================\n");
        
        // Basic generation tests
        testBasicGeneration();
        testNumericMode();
        testAlphanumericMode();
        testByteMode();
        testErrorCorrectionLevels();
        testTextOutput();
        testAsciiOutput();
        testImageOutput();
        testDifferentSizes();
        testCapacity();
        testEdgeCases();
        
        // Summary
        System.out.println("\n========================================");
        System.out.println("Test Summary:");
        System.out.println("  Passed: " + testsPassed);
        System.out.println("  Failed: " + testsFailed);
        System.out.println("  Total:  " + (testsPassed + testsFailed));
        System.out.println("========================================");
        
        if (testsFailed > 0) {
            System.exit(1);
        }
    }
    
    private static void testBasicGeneration() {
        System.out.println("Testing basic QR code generation...");
        
        try {
            // Test basic generation
            mod.QRCode qr = mod.generate("Hello, World!");
            assertNotNull("QR code should not be null", qr);
            assertTrue("QR code size should be positive", qr.size > 0);
            assertNotNull("QR code modules should not be null", qr.modules);
            assertEquals("QR code data should match input", "Hello, World!", qr.data);
            assertNotNull("QR code error correction level should not be null", qr.errorCorrectionLevel);
            assertNotNull("QR code mode should not be null", qr.mode);
            
            // Test that modules array has correct dimensions
            assertEquals("Modules array height should match size", qr.size, qr.modules.length);
            assertEquals("Modules array width should match size", qr.size, qr.modules[0].length);
            
            pass("Basic generation test");
        } catch (Exception e) {
            fail("Basic generation test", e.getMessage());
        }
    }
    
    private static void testNumericMode() {
        System.out.println("Testing numeric mode encoding...");
        
        try {
            // Test pure numeric data
            mod.QRCode qr = mod.generate("1234567890");
            assertEquals("Numeric data should use numeric mode", mod.Mode.NUMERIC, qr.mode);
            
            // Test longer numeric string
            qr = mod.generate("12345678901234567890");
            assertEquals("Long numeric data should use numeric mode", mod.Mode.NUMERIC, qr.mode);
            
            pass("Numeric mode test");
        } catch (Exception e) {
            fail("Numeric mode test", e.getMessage());
        }
    }
    
    private static void testAlphanumericMode() {
        System.out.println("Testing alphanumeric mode encoding...");
        
        try {
            // Test alphanumeric data
            mod.QRCode qr = mod.generate("HELLO WORLD");
            assertEquals("Alphanumeric data should use alphanumeric mode", mod.Mode.ALPHANUMERIC, qr.mode);
            
            // Test with special characters
            qr = mod.generate("ABC123 $%*+-./:");
            assertEquals("Alphanumeric with special chars should use alphanumeric mode", mod.Mode.ALPHANUMERIC, qr.mode);
            
            pass("Alphanumeric mode test");
        } catch (Exception e) {
            fail("Alphanumeric mode test", e.getMessage());
        }
    }
    
    private static void testByteMode() {
        System.out.println("Testing byte mode encoding...");
        
        try {
            // Test byte data (lowercase letters)
            mod.QRCode qr = mod.generate("hello world");
            assertEquals("Lowercase data should use byte mode", mod.Mode.BYTE, qr.mode);
            
            // Test with Unicode
            qr = mod.generate("Hello, World!");
            assertEquals("ASCII with punctuation should use byte mode", mod.Mode.BYTE, qr.mode);
            
            // Test with mixed content
            qr = mod.generate("Test123!@#");
            assertEquals("Mixed content should use byte mode", mod.Mode.BYTE, qr.mode);
            
            pass("Byte mode test");
        } catch (Exception e) {
            fail("Byte mode test", e.getMessage());
        }
    }
    
    private static void testErrorCorrectionLevels() {
        System.out.println("Testing error correction levels...");
        
        try {
            String data = "Hello, World!";
            
            // Test all error correction levels
            mod.QRCode qrL = mod.generate(data, mod.ErrorCorrectionLevel.L);
            assertEquals("Should use L error correction", mod.ErrorCorrectionLevel.L, qrL.errorCorrectionLevel);
            
            mod.QRCode qrM = mod.generate(data, mod.ErrorCorrectionLevel.M);
            assertEquals("Should use M error correction", mod.ErrorCorrectionLevel.M, qrM.errorCorrectionLevel);
            
            mod.QRCode qrQ = mod.generate(data, mod.ErrorCorrectionLevel.Q);
            assertEquals("Should use Q error correction", mod.ErrorCorrectionLevel.Q, qrQ.errorCorrectionLevel);
            
            mod.QRCode qrH = mod.generate(data, mod.ErrorCorrectionLevel.H);
            assertEquals("Should use H error correction", mod.ErrorCorrectionLevel.H, qrH.errorCorrectionLevel);
            
            pass("Error correction levels test");
        } catch (Exception e) {
            fail("Error correction levels test", e.getMessage());
        }
    }
    
    private static void testTextOutput() {
        System.out.println("Testing text output...");
        
        try {
            mod.QRCode qr = mod.generate("TEST");
            String text = qr.toText();
            
            assertNotNull("Text output should not be null", text);
            assertTrue("Text output should not be empty", text.length() > 0);
            assertTrue("Text output should contain block characters", text.contains("\u2588"));
            
            // Test generateText convenience method
            String directText = mod.generateText("TEST");
            assertNotNull("generateText should return non-null", directText);
            assertTrue("generateText should return non-empty", directText.length() > 0);
            
            pass("Text output test");
        } catch (Exception e) {
            fail("Text output test", e.getMessage());
        }
    }
    
    private static void testAsciiOutput() {
        System.out.println("Testing ASCII output...");
        
        try {
            mod.QRCode qr = mod.generate("TEST");
            String ascii = qr.toAscii();
            
            assertNotNull("ASCII output should not be null", ascii);
            assertTrue("ASCII output should not be empty", ascii.length() > 0);
            assertTrue("ASCII output should contain # characters", ascii.contains("#"));
            
            // Test generateAscii convenience method
            String directAscii = mod.generateAscii("TEST");
            assertNotNull("generateAscii should return non-null", directAscii);
            assertTrue("generateAscii should return non-empty", directAscii.length() > 0);
            
            pass("ASCII output test");
        } catch (Exception e) {
            fail("ASCII output test", e.getMessage());
        }
    }
    
    private static void testImageOutput() {
        System.out.println("Testing image output...");
        
        try {
            mod.QRCode qr = mod.generate("TEST");
            BufferedImage image = qr.toImage();
            
            assertNotNull("Image should not be null", image);
            assertTrue("Image width should be positive", image.getWidth() > 0);
            assertTrue("Image height should be positive", image.getHeight() > 0);
            assertEquals("Image should be square", image.getWidth(), image.getHeight());
            
            // Test with custom module size
            BufferedImage largeImage = qr.toImage(8, 4);
            assertTrue("Large image should be bigger", largeImage.getWidth() > image.getWidth());
            
            // Test generateImage convenience method
            BufferedImage directImage = mod.generateImage("TEST");
            assertNotNull("generateImage should return non-null", directImage);
            
            pass("Image output test");
        } catch (Exception e) {
            fail("Image output test", e.getMessage());
        }
    }
    
    private static void testDifferentSizes() {
        System.out.println("Testing different data sizes...");
        
        try {
            // Small data
            mod.QRCode qrSmall = mod.generate("A");
            assertTrue("Small data should generate valid QR", qrSmall.size > 0);
            
            // Medium data
            mod.QRCode qrMedium = mod.generate("Hello, World! This is a test.");
            assertTrue("Medium data should generate valid QR", qrMedium.size > 0);
            
            // Larger data
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < 100; i++) {
                sb.append("TEST");
            }
            mod.QRCode qrLarge = mod.generate(sb.toString());
            assertTrue("Large data should generate valid QR", qrLarge.size > 0);
            
            pass("Different sizes test");
        } catch (Exception e) {
            fail("Different sizes test", e.getMessage());
        }
    }
    
    private static void testCapacity() {
        System.out.println("Testing capacity functions...");
        
        try {
            // Test getQrSize
            int size1 = mod.getQrSize(1);
            assertEquals("Version 1 should be 21x21", 21, size1);
            
            int size2 = mod.getQrSize(2);
            assertEquals("Version 2 should be 25x25", 25, size2);
            
            int size40 = mod.getQrSize(40);
            assertEquals("Version 40 should be 177x177", 177, size40);
            
            // Test getCapacity
            int capacity = mod.getCapacity(1, mod.Mode.BYTE, mod.ErrorCorrectionLevel.M);
            assertTrue("Capacity should be positive", capacity > 0);
            
            // Test getMinimumVersion
            int minVersion = mod.getMinimumVersion("HELLO", mod.ErrorCorrectionLevel.M);
            assertTrue("Minimum version should be at least 1", minVersion >= 1);
            assertTrue("Minimum version should be at most 40", minVersion <= 40);
            
            pass("Capacity test");
        } catch (Exception e) {
            fail("Capacity test", e.getMessage());
        }
    }
    
    private static void testEdgeCases() {
        System.out.println("Testing edge cases...");
        
        try {
            // Test empty string (should throw exception)
            try {
                mod.generate("");
                fail("Empty string test", "Should throw exception for empty string");
            } catch (IllegalArgumentException e) {
                // Expected
            }
            
            // Test null (should throw exception)
            try {
                mod.generate(null);
                fail("Null test", "Should throw exception for null");
            } catch (IllegalArgumentException e) {
                // Expected
            }
            
            // Test canEncode
            assertTrue("canEncode should return true for valid data", 
                      mod.canEncode("TEST", mod.ErrorCorrectionLevel.M));
            
            // Test module access
            mod.QRCode qr = mod.generate("TEST");
            boolean module = qr.getModule(0, 0);
            // Just verify it doesn't throw
            
            // Test out of bounds access
            boolean outOfBounds = qr.getModule(-1, -1);
            assertFalse("Out of bounds should return false", outOfBounds);
            
            pass("Edge cases test");
        } catch (Exception e) {
            fail("Edge cases test", e.getMessage());
        }
    }
    
    // ==================== Assertion Helpers ====================
    
    private static void assertTrue(String message, boolean condition) {
        if (!condition) {
            throw new AssertionError(message);
        }
    }
    
    private static void assertFalse(String message, boolean condition) {
        if (condition) {
            throw new AssertionError(message);
        }
    }
    
    private static void assertEquals(String message, Object expected, Object actual) {
        if (expected == null ? actual != null : !expected.equals(actual)) {
            throw new AssertionError(message + " - Expected: " + expected + ", Actual: " + actual);
        }
    }
    
    private static void assertNotNull(String message, Object object) {
        if (object == null) {
            throw new AssertionError(message);
        }
    }
    
    private static void pass(String testName) {
        testsPassed++;
        System.out.println("  [PASS] " + testName);
    }
    
    private static void fail(String testName, String reason) {
        testsFailed++;
        System.out.println("  [FAIL] " + testName + ": " + reason);
    }
}
